lib/ocpp/v201/init_device_model_db.cpp

```diff
@@ -0,0 +1,1551 @@
+}
+
+InitDeviceModelDb::~InitDeviceModelDb() {
+    close_connection();
+}
+
+void InitDeviceModelDb::initialize_database(const std::filesystem::path& schemas_path,
+                                            bool delete_db_if_exists = true) {
+    execute_init_sql(delete_db_if_exists);
+
+    // Get existing EVSE and Connector components from the database.
+    std::vector<ComponentKey> existing_components;
+    DeviceModelMap device_model;
+    if (this->database_exists) {
+        try {
+            if (this->database->get_user_version() == 0) {
+                EVLOG_AND_THROW(
+                    InitDeviceModelDbError("Database does not support migrations yet, please update the database."));
+            }
+        } catch (const std::runtime_error& e) {
+            EVLOG_AND_THROW(
+                InitDeviceModelDbError("Database does not support migrations yet, please update the database."));
+        }
+
+        existing_components = get_all_connector_and_evse_components_fom_db();
+    }
+
+    // Get component schemas from the filesystem.
+    std::map<ComponentKey, std::vector<DeviceModelVariable>> component_schemas =
+        get_all_component_schemas(schemas_path);
+
+    // Remove components from db if they do not exist in the component schemas
+    if (this->database_exists) {
+        remove_not_existing_components_from_db(component_schemas, existing_components);
+    }
+
+    // Starting a transaction makes this a lot faster (inserting all components takes a few seconds without it and a
+    // few milliseconds if it is done inside a transaction).
+    std::unique_ptr<common::DatabaseTransactionInterface> transaction = database->begin_transaction();
+    insert_components(component_schemas, existing_components);
+    transaction->commit();
+}
+
+bool InitDeviceModelDb::insert_config_and_default_values(const std::filesystem::path& schemas_path,
+                                                         const std::filesystem::path& config_path) {
+    bool success = true;
+    std::map<ComponentKey, std::vector<DeviceModelVariable>> components = get_all_components_from_db();
+
+    std::map<ComponentKey, std::vector<VariableAttributeKey>> default_values =
+        get_component_default_values(schemas_path);
+    std::map<ComponentKey, std::vector<VariableAttributeKey>> config_values = get_config_values(config_path);
+
+    const std::vector<std::string> errors = check_config_integrity(components, config_values);
+
+    if (!errors.empty()) {
+        const std::string output = std::accumulate(++errors.begin(), errors.end(), *errors.begin(),
+                                                   [](auto& a, auto& b) { return a + "\n" + b; });
+        EVLOG_AND_THROW(
+            InitDeviceModelDbError("Config not consistent with device model component schema's: \n" + output));
+    }
+
+    std::unique_ptr<common::DatabaseTransactionInterface> transaction = database->begin_transaction();
```

Comment:
Starting a transaction here and ending it at the end of the function. If anything throws in between, transaction will not be committed. I think that's the idea but I can of course also add a try/catch there. What do you think?